[visionOS] Add a mechanism to check user authorization for gamepads

achan00 opened the pull request:
#### 015450fc9083884904de9d26c87319b6b123f27f
<pre>
[visionOS] Add a mechanism to check user authorization for gamepads
<a href="https://bugs.webkit.org/show_bug.cgi?id=276842">https://bugs.webkit.org/show_bug.cgi?id=276842</a>
<a href="https://rdar.apple.com/132103378">rdar://132103378</a>

Reviewed by Brady Eidson.

Add a new gamepadAccessRequiresExplicitConsent member in page configuration
to control whether gamepad access is blocked from webpage until
-[WKWebView _setAllowGamepadsAccess:YES] is called. By default,
gamepadAccessRequiresExplicitConsent is false.

Update the logic in GamepadManager that determines when to reveal gamepads
to navigators/windows that are just registered to also consult the
page&apos;s m_gamepadAccessGranted boolean. If Page::gamepadAccessGranted() returns
false when the navigator/window is first registered with the manager, they
are first placed in a separate quarantined set, and they don&apos;t automatically
get gamepads revealed to them on new gamepad connection or gamepad activity.
They are removed from quarantine only after their pages have been explicitly
granted gamepad access.

Also, add WKUIDelegatePrivate methods to communicate gamepad
access/connection state changes.

* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::GamepadManager::findUnquarantinedNavigatorsAndWindows):
Gather all navigators/windows that are not waiting for explicit
user consent for gamepad access.
(WebCore::GamepadManager::platformGamepadConnected):
Do not notify quarantined navigators/windows of new gamepad connections.
(WebCore::GamepadManager::platformGamepadDisconnected):
Ditto for gamepad disconnections.
(WebCore::GamepadManager::registerNavigator):
Add to the quarantined set if access has not been granted.
(WebCore::GamepadManager::unregisterNavigator):
(WebCore::GamepadManager::registerDOMWindow):
Ditto for window.
(WebCore::GamepadManager::unregisterDOMWindow):
(WebCore::GamepadManager::updateQuarantineStatus):
This is called when a page&apos;s gamepad access state has changed.
It goes through the quarantined sets to see which ones are from
pages that have been granted access. Promote them to the blind
sets and remove them from the quarantined sets.
* Source/WebCore/Modules/gamepad/GamepadManager.h:
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::protectedPage const):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::m_activeNowPlayingSessionUpdateTimer):
(WebCore::Page::didCommitLoad):
Reset gamepad access on new page load.
(WebCore::Page::allowGamepadAccess):
If gamepad access state has changed, tell GamepadManager to update its
quarantined sets of navigators/windows.
(WebCore::Page::initializeGamepadAccessForPageLoad):
* Source/WebCore/page/Page.h:
(WebCore::Page::gamepadAccessGranted const):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/gamepad/ShouldRequireExplicitConsentForGamepadAccess.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::gamepadAccessRequiresExplicitConsent const):
(API::PageConfiguration::setGamepadAccessRequiresExplicitConsent):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _gamepadsConnected]):
(-[WKWebView _gamepadsConnectedStateChanged]):
(-[WKWebView _setGamepadsRecentlyAccessed:]):
(-[WKWebView _setAllowGamepadsAccess]):
(-[WKWebView _setAllowGamepadsInput:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]):
(-[WKWebViewConfiguration initWithCoder:]):
(-[WKWebViewConfiguration _gamepadAccessRequiresExplicitConsent]):
(-[WKWebViewConfiguration _setGamepadAccessRequiresExplicitConsent:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::gamepadsConnectedStateChanged):
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.cpp:
(WebKit::UIGamepadProvider::gamepadSyncTimerFired):
Update the new active page&apos;s gamepads connection state.
(WebKit::UIGamepadProvider::isAnyGamepadConnected const):
Check if there&apos;s any non-null gamepad in m_gamepads.
(WebKit::UIGamepadProvider::viewBecameActive):
Update the new active page&apos;s gamepads connection state.
(WebKit::UIGamepadProvider::viewBecameInactive):
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::gamepadsConnectedStateChanged):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
Make sure we update the recent gamepad access state to false on page navigation.
(WebKit::WebPageProxy::resetStateAfterProcessExited):
Ditto.
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::resetRecentGamepadAccessState):
Besides cancelling m_recentGamepadAccessHysteresis, we need to make sure
the clients know the recent gamepad access state has been reset to false.
(WebKit::WebPageProxy::setGamepadsConnected):
Notify page client if the gamepad connection state changes.
(WebKit::WebPageProxy::allowGamepadAccess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::allowGamepadAccess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/281432@main">https://commits.webkit.org/281432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1136662c1e3bc1f96eecb3113772d63c5b8869be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63805 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10419 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46888 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10590 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/7274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36589 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51853 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/29391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9335 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55219 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9373 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65535 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3816 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9234 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55894 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51834 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/56037 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13273 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3167 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35049 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebCore/platform/gamepad/ShouldRequireExplicitConsentForGamepadAccess.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace WebCore {
+
+enum class ShouldRequireExplicitConsentForGamepadAccess : bool { No, Yes };
```
Overkill to have put this in its own header ;)
(But you can leave it)

achan00 committed:
[visionOS] Add a mechanism to check user authorization for gamepads
https://bugs.webkit.org/show_bug.cgi?id=276842
rdar://132103378

Reviewed by Brady Eidson.

Add a new gamepadAccessRequiresExplicitConsent member in page configuration
to control whether gamepad access is blocked from webpage until
-[WKWebView _setAllowGamepadsAccess:YES] is called. By default,
gamepadAccessRequiresExplicitConsent is false.

Update the logic in GamepadManager that determines when to reveal gamepads
to navigators/windows that are just registered to also consult the
page's m_gamepadAccessGranted boolean. If Page::gamepadAccessGranted() returns
false when the navigator/window is first registered with the manager, they
are first placed in a separate quarantined set, and they don't automatically
get gamepads revealed to them on new gamepad connection or gamepad activity.
They are removed from quarantine only after their pages have been explicitly
granted gamepad access.

Also, add WKUIDelegatePrivate methods to communicate gamepad
access/connection state changes.

* Source/WebCore/Modules/gamepad/GamepadManager.cpp:
(WebCore::GamepadManager::findUnquarantinedNavigatorsAndWindows):
Gather all navigators/windows that are not waiting for explicit
user consent for gamepad access.
(WebCore::GamepadManager::platformGamepadConnected):
Do not notify quarantined navigators/windows of new gamepad connections.
(WebCore::GamepadManager::platformGamepadDisconnected):
Ditto for gamepad disconnections.
(WebCore::GamepadManager::registerNavigator):
Add to the quarantined set if access has not been granted.
(WebCore::GamepadManager::unregisterNavigator):
(WebCore::GamepadManager::registerDOMWindow):
Ditto for window.
(WebCore::GamepadManager::unregisterDOMWindow):
(WebCore::GamepadManager::updateQuarantineStatus):
This is called when a page's gamepad access state has changed.
It goes through the quarantined sets to see which ones are from
pages that have been granted access. Promote them to the blind
sets and remove them from the quarantined sets.
* Source/WebCore/Modules/gamepad/GamepadManager.h:
* Source/WebCore/Modules/gamepad/NavigatorGamepad.cpp:
(WebCore::NavigatorGamepad::protectedPage const):
* Source/WebCore/Modules/gamepad/NavigatorGamepad.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::m_activeNowPlayingSessionUpdateTimer):
(WebCore::Page::didCommitLoad):
Reset gamepad access on new page load.
(WebCore::Page::allowGamepadAccess):
If gamepad access state has changed, tell GamepadManager to update its
quarantined sets of navigators/windows.
(WebCore::Page::initializeGamepadAccessForPageLoad):
* Source/WebCore/page/Page.h:
(WebCore::Page::gamepadAccessGranted const):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/gamepad/ShouldRequireExplicitConsentForGamepadAccess.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::gamepadAccessRequiresExplicitConsent const):
(API::PageConfiguration::setGamepadAccessRequiresExplicitConsent):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _gamepadsConnected]):
(-[WKWebView _gamepadsConnectedStateChanged]):
(-[WKWebView _setGamepadsRecentlyAccessed:]):
(-[WKWebView _setAllowGamepadsAccess]):
(-[WKWebView _setAllowGamepadsInput:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]):
(-[WKWebViewConfiguration initWithCoder:]):
(-[WKWebViewConfiguration _gamepadAccessRequiresExplicitConsent]):
(-[WKWebViewConfiguration _setGamepadAccessRequiresExplicitConsent:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::gamepadsConnectedStateChanged):
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.cpp:
(WebKit::UIGamepadProvider::gamepadSyncTimerFired):
Update the new active page's gamepads connection state.
(WebKit::UIGamepadProvider::isAnyGamepadConnected const):
Check if there's any non-null gamepad in m_gamepads.
(WebKit::UIGamepadProvider::viewBecameActive):
Update the new active page's gamepads connection state.
(WebKit::UIGamepadProvider::viewBecameInactive):
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.h:
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::gamepadsConnectedStateChanged):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
Make sure we update the recent gamepad access state to false on page navigation.
(WebKit::WebPageProxy::resetStateAfterProcessExited):
Ditto.
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::resetRecentGamepadAccessState):
Besides cancelling m_recentGamepadAccessHysteresis, we need to make sure
the clients know the recent gamepad access state has been reset to false.
(WebKit::WebPageProxy::setGamepadsConnected):
Notify page client if the gamepad connection state changes.
(WebKit::WebPageProxy::allowGamepadAccess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::allowGamepadAccess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/281432@main
